packages/preview-service/services/screenshot.ts

```diff
@@ -0,0 +1,120 @@
+import { reduce } from 'lodash'
+import {
+  shouldBeHeadless,
+  getPuppeteerUserDataDir,
+  getChromiumExecutablePath
+} from '../utils/env'
+import type { Logger } from 'pino'
+import { PuppeteerClientInterface } from '../clients/puppeteer'
+import type { ObjectIdentifier } from '../domain/domain'
+
+export type GetScreenshot = (
+  params: ObjectIdentifier
+) => Promise<{ [key: string]: string } | null>
+
+export const getScreenshotFactory =
+  (deps: {
+    puppeteerClient: PuppeteerClientInterface
+    logger: Logger
+    serviceOrigin: string
+  }): GetScreenshot =>
+  async (params) => {
+    const objectUrl = `${deps.serviceOrigin}/streams/${params.streamId}/objects/${params.objectId}`
+    const launchParams = {
+      //TODO as the launch params are touching environment variables, should they be dependency injected?
+      headless: shouldBeHeadless(),
+      userDataDir: getPuppeteerUserDataDir(),
+      executablePath: getChromiumExecutablePath(),
+      protocolTimeout: 3600_000, //TODO make this configurable to match the maximum timeout of the service
+      // we trust the web content that is running, so can disable the sandbox
+      // disabling the sandbox allows us to run the docker image without linux kernel privileges
+      args: ['--no-sandbox', '--disable-setuid-sandbox', '--disable-dev-shm-usage']
+    }
+
+    await deps.puppeteerClient.init(launchParams)
```

Comment:
It bothers me a little bit that PuppeterClient creation & initialization are disconnected and done in different files. Makes it possible to work with a PuppeteerClient that hasn't been initialized without knowing it.

tbh, I'd convert PuppeteerClient away from a class into an async function, to achieve the equivalent of an "async constructor", that way you wouldn't have to worry about passing around a non-initialized client instance

```
const createPuppeteerClient = async (...initParams) => {
  return {
    loadPageAndEvaluateScript,
    dispose
  }
}
```

Also - what if the client is already initialized? It seems that init() is not idempotent and can be ran multiple times, creating a new browser each time and apparently not cleaning the old one up

Reply:
Maybe you can achieve async initialization with classes too? I don't know enough about them to say, but in other programming languages you could have a private constructor and a public async static function that would be the only way to initialize the class.

The functional approach seems less complicated to me imo